README
NAME
    Plack::App::Proxy - proxy requests

SYNOPSIS
      use Plack::Builder;

      # proxy all requests for /static to 127.0.0.1:80
      builder {
          mount "/static" => Plack::App::Proxy->new(remote => "http://127.0.0.1:80")->to_app;
      };

      # Call from other app
      my $proxy = Plack::App::Proxy->new->to_app;
      my $app = sub {
          my $env = shift;
          ...
          $env->{'plack.proxy.url'} = $url;
          $proxy->($env);
      };

DESCRIPTION
    Plack::App::Proxy is a middleware-aware proxy application for Plack.

OPTIONS
    remote
          Plack::App::Proxy->new(remote => 'http://perl.org')->to_app;

        Specifies the base remote URL to proxy requests to.

          builder {
              mount "/example",
                  Plack::App::Proxy->new(remote => 'http://example.com/app/foo')->to_app;
          };

        This proxies incoming requests for "/example/bar" proxied to
        "http://example.com/app/foo/bar".

    preserve_host_header
        Preserves the original Host header, which is useful when you do
        reverse proxying to the internal hosts.

    backend
        The HTTP backend to use. This dist comes with "LWP" and
        "AnyEvent::HTTP" backends. "AnyEvent::HTTP" is the default if no
        backend is specified.

    options
        The options for the HTTP backend instance.

MIDDLEWARE CONFIGURATIONS
    This application is just like a normal PSGI application and is
    middleware aware, which means you can modify proxy requests (and
    responses) using Plack middleware stack.

    It also supports the following special environment variables:

    plack.proxy.url
        Overrides the proxy request URL.

    plack.proxy.remote
        Overrides the base URL path to proxy to.

    For example, the following builder code allows you to proxy all GET
    requests for .png paths to the lolcat image (yes, a silly example) but
    proxies to the internal host otherwise.

      my $mw = sub {
          my $app = shift;
          sub {
              my $env = shift;
              if ($env->{REQUEST_METHOD} eq 'GET' && $env->{PATH_INFO} =~ /\.png$/) {
                  $env->{'plack.proxy.url'} = 'http://lolcat.example.com/lol.png';
              }
              $app->($env);
          };
      };

      use Plack::Builder;

      builder {
          enable $mw;
          Plack::App::Proxy->new(remote => 'http://10.0.0.1:8080')->to_app;
      };

AUTHOR
    Lee Aylward

    Masahiro Honma

    Tatsuhiko Miyagawa

    Jesse Luehrs

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    Plack::Builder



Changes
Revision history for Plack-App-Proxy

0.29    08/31/2013
        - Better cookie support with the AnyEvent::HTTP backend (Manni Heumann)

0.28    08/24/2013
        - allow passing options to backend (clkao)

0.27    02/26/2013
        - Remove more HTTP::Headers assumptions

0.26    02/26/2013
        - Add parent dependency

0.25    11/10/2012
        - Fix problem with META.yml

0.24    11/10/2012
        - Skip CONNECT test on Windows

0.23    11/05/2012
        - Fix test failures from newer HTTP::Headers

0.22
        - Filter out Status header since it's forbidden in PSGI spec (miyagawa)

0.21
        - Bump version to fix some CPAN issues.

0.20
        - Fix auto guessing of RewriteLocation.

0.19
        - Add pluggable backends, since AnyEvent seems to be flaky on
          some systems. (doy)

0.18    02/17/2010
        - Fixes for RewriteLocation when used on a mounted proxy
          (hiratara, rbuels)
        - Ignore $ENV{http_proxy}
        - Include reason for failure in 502 response
        - Disabled persistent HTTP connection option in AnyEvent::HTTP
          
0.17    01/19/2010
        - Bump version to fix some CPAN issues

0.16    01/12/2010
        - Rewrite proxy code to use AnyEvent::HTTP's on_body callback.
          This gets chunked and gzipped responses working.

0.15    03/26/2010
        - Added the LoadBalancer middleware (franck cuny)
        - Fixed SYNOPSIS of RewriteLocation (franck cuny)

0.14    03/12/2010
        - Fixed META.yml broken (Thanks to miyagawa)
        - Removed dependencies on Plack-Request

0.13    02/27/2010
        - Added Connect middleware component
        - Fixed the problem that responses which status code isn't 200
          caused freeze on blocking servers
        - Fixed the leak of $env with perl 5.8.x
        - Relocated test programs

0.12    02/09/2010
        - Fixed the pod syntax
        - Had tests pass without Plack::Server::AnyEvent

0.11    02/08/2010
        - INCOMPATIBLE: the config option 'host' is now renamed to 'remote'
        - Made sure the code forward compatible with Plack 1.0
        - Added a PSGI environment vars support to write middleware
        - callback options for url/host is deprecated. See the docs how to do
          that using middleware
        - Fixed a bug where Location header was rewritten
        - Added RewriteLocation and AddVia middleware components

0.10    01/29/2010
        Fixed the bug which drops QUERY_STRING from the original request.
        Enabled the Set-Cookies headers in the response.
        (Any other headers are also enabled. See RFC 2616.)

0.09    01/13/2010
        The tests no longer depend on the Google site.

0.08    01/09/2010
        Fixed memory leaks for perl 5.8.x.
        Added more tests.

0.07    01/05/2010
        Only use AnyEvent::HTTP for both blocking and non-blocking
        servers

0.06    01/04/2010
        Fixed typo in MANIFEST.SKIP

0.05    01/04/2010
        Re-upload with missing changes

0.04    01/04/2010
        Added gzip support (hiratara)
        AnyEvent::HTTP fixes (hiratara)

0.03    12/30/2009
        Don't use google.com for tests because of localizations

0.02    12/25/2009
        Forgot to include Makefile.PL

0.01    12/22/2009
        First version, released on an unsuspecting world.



lib/Plack/App/Proxy.pm
package Plack::App::Proxy;

use strict;
use 5.008_001;
use parent 'Plack::Component';
use Plack::Util::Accessor qw/remote preserve_host_header backend options/;
use Plack::Request;
use Plack::Util;
use HTTP::Headers;

our $VERSION = '0.29';

sub prepare_app {
    my $self = shift;
    $self->backend($ENV{PLACK_PROXY_BACKEND} || 'AnyEvent::HTTP') unless defined $self->backend;
}

# hop-by-hop headers (see also RFC2616)
my @hop_by_hop = qw(
    Connection Keep-Alive Proxy-Authenticate Proxy-Authorization
    TE Trailer Transfer-Encoding Upgrade
);

sub filter_headers {
    my $self = shift;
    my ( $headers ) = @_;

    # Save from known hop-by-hop deletion.
    my @connection_tokens = $headers->header('Connection');

    # Remove hop-by-hop headers.
    $headers->remove_header( $_ ) for @hop_by_hop;

    # Connection header's tokens are also hop-by-hop.
    for my $token ( @connection_tokens ){
        $headers->remove_header( $_ ) for split /\s*,\s*/, $token;
    }
}

sub build_url_from_env {
    my($self, $env) = @_;

    return $env->{'plack.proxy.url'}
        if exists $env->{'plack.proxy.url'};

    my $url = $env->{'plack.proxy.remote'} || $self->remote
        or return;

    # avoid double slashes
    $url =~ s!/$!! unless $env->{SCRIPT_NAME} && $env->{SCRIPT_NAME} =~ m!/$!;

    $url .= $env->{PATH_INFO} || '';
    $url .= '?' . $env->{QUERY_STRING} if defined $env->{QUERY_STRING} && length $env->{QUERY_STRING} > 0;

    return $url;
}

sub build_headers_from_env {
    my($self, $env, $req) = @_;

    my $headers = $req->headers->clone;
    $headers->header("X-Forwarded-For" => $env->{REMOTE_ADDR});
    $headers->remove_header("Host") unless $self->preserve_host_header;
    $self->filter_headers( $headers );

    +{ map {$_ => scalar $headers->header($_) } $headers->header_field_names };
}

sub call {
    my ($self, $env) = @_;

    unless ($env->{'psgi.streaming'}) {
        die "Plack::App::Proxy only runs with the server with psgi.streaming support";
    }

    my $url = $self->build_url_from_env($env)
        or return [502, ["Content-Type","text/html"], ["Can't determine proxy remote URL"]];

    my $req = Plack::Request->new($env);
    my $headers = $self->build_headers_from_env($env, $req);

    my $method  = $env->{REQUEST_METHOD};
    my $content = $req->content;

    my $backend_class = Plack::Util::load_class(
        $self->backend, 'Plack::App::Proxy::Backend'
    );

    return $backend_class->new(
        url              => $url,
        req              => $req,
        headers          => $headers,
        method           => $method,
        content          => $content,
        options          => $self->options,
        response_headers => sub { $self->response_headers(@_) },
    )->call($env);
}

sub response_headers {
    my ($self, $headers) = @_;

    $self->filter_headers( $headers );

    # Remove PSGI forbidden headers
    $headers->remove_header('Status');

    my @headers;
    $headers->scan( sub { push @headers, @_ } );

    return @headers;
}

1;

__END__

=head1 NAME

Plack::App::Proxy - proxy requests

=head1 SYNOPSIS

  use Plack::Builder;

  # proxy all requests for /static to 127.0.0.1:80
  builder {
      mount "/static" => Plack::App::Proxy->new(remote => "http://127.0.0.1:80")->to_app;
  };

  # Call from other app
  my $proxy = Plack::App::Proxy->new->to_app;
  my $app = sub {
      my $env = shift;
      ...
      $env->{'plack.proxy.url'} = $url;
      $proxy->($env);
  };

=head1 DESCRIPTION

Plack::App::Proxy is a middleware-aware proxy application for Plack.

=head1 OPTIONS

=over 4

=item remote

  Plack::App::Proxy->new(remote => 'http://perl.org')->to_app;

Specifies the base remote URL to proxy requests to.

  builder {
      mount "/example",
          Plack::App::Proxy->new(remote => 'http://example.com/app/foo')->to_app;
  };

This proxies incoming requests for C</example/bar> proxied to
C<http://example.com/app/foo/bar>.

=item preserve_host_header

Preserves the original Host header, which is useful when you do
reverse proxying to the internal hosts.

=item backend

The HTTP backend to use. This dist comes with C<LWP> and C<AnyEvent::HTTP>
backends. C<AnyEvent::HTTP> is the default if no backend is specified.

=item options

The options for the HTTP backend instance.

=back

=head1 MIDDLEWARE CONFIGURATIONS

This application is just like a normal PSGI application and is
middleware aware, which means you can modify proxy requests (and
responses) using Plack middleware stack.

It also supports the following special environment variables:

=over 4

=item plack.proxy.url

Overrides the proxy request URL.

=item plack.proxy.remote

Overrides the base URL path to proxy to.

=back

For example, the following builder code allows you to proxy all GET
requests for .png paths to the lolcat image (yes, a silly example) but
proxies to the internal host otherwise.

  my $mw = sub {
      my $app = shift;
      sub {
          my $env = shift;
          if ($env->{REQUEST_METHOD} eq 'GET' && $env->{PATH_INFO} =~ /\.png$/) {
              $env->{'plack.proxy.url'} = 'http://lolcat.example.com/lol.png';
          }
          $app->($env);
      };
  };

  use Plack::Builder;

  builder {
      enable $mw;
      Plack::App::Proxy->new(remote => 'http://10.0.0.1:8080')->to_app;
  };

=head1 AUTHOR

Lee Aylward

Masahiro Honma

Tatsuhiko Miyagawa

Jesse Luehrs

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<Plack::Builder>

=cut
